auth/oidc.go

```diff
@@ -439,3 +533,180 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
 	uri.RawQuery = rawQuery.Encode()
 	return uri.String(), nil
 }
+
+// StartDiscoverySync runs the provider metadata discovery sync task in background.
+func (op *OIDCProvider) StartDiscoverySync(discoveryURL string) error {
+	op.stopDiscoverySync = make(chan struct{})
+	duration, err := op.runDiscoverySync(discoveryURL)
+	if err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(duration):
+				duration, err = op.runDiscoverySync(discoveryURL)
+				if err != nil {
+					base.Warnf("OpenID Connect provider discovery sync ends up in error: %v, next retry in %v", err, duration)
+				}
+			case <-op.stopDiscoverySync:
+				base.Debugf(base.KeyAll, "Terminating OpenID Connect provider discovery sync")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// runDiscoverySync polls the provider metadata from discovery endpoint and refresh the ID token verifier.
+func (op *OIDCProvider) runDiscoverySync(discoveryURL string) (time.Duration, error) {
+	base.Debugf(base.KeyAuth, "Syncing provider metadata from: %s", base.UD(discoveryURL))
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	var expiresAt time.Time
+	var ttl time.Duration
+	var ok bool
+	ttl, ok, err = cacheable(resp.Header)
```

Comment:
Here you get the ttl as a duration, then add it to time.Now() to get expiresAt.  Then you pass expiresAt to nextSyncAfter, which turns it back into a duration via expiresAt.Sub(time.Now()).  Seems like you should refactor to just use time.Duration throughout.